x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -72,50 +82,140 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
+                    listener.onFailure(new NoNodeAvailableException("no nodes have enough disk space to hold one copy of the index [" +
+                        indexName + "] onto prior to shrink"));
+                    return;
+                }
+
+                Optional<String> nodeId = selectSingleNode(validNodeIds, nodeShardsStorageBytes);
+
+                if (nodeId.isPresent()) {
+                    Settings settings = Settings.builder()
                         .put(IndexMetadata.INDEX_ROUTING_REQUIRE_GROUP_SETTING.getKey() + "_id", nodeId.get())
                         .putNull(ShardsLimitAllocationDecider.INDEX_TOTAL_SHARDS_PER_NODE_SETTING.getKey()).build();
-                UpdateSettingsRequest updateSettingsRequest = new UpdateSettingsRequest(indexName)
+                    UpdateSettingsRequest updateSettingsRequest = new UpdateSettingsRequest(indexName)
                         .masterNodeTimeout(TimeValue.MAX_VALUE)
                         .settings(settings);
-                getClient().admin().indices().updateSettings(updateSettingsRequest,
+                    getClient().admin().indices().updateSettings(updateSettingsRequest,
                         ActionListener.wrap(response -> listener.onResponse(null), listener::onFailure));
-            } else {
-                // No nodes currently match the allocation rules, so report this as an error and we'll retry
-                logger.debug("could not find any nodes to allocate index [{}] onto prior to shrink", indexName);
-                listener.onFailure(new NoNodeAvailableException("could not find any nodes to allocate index [" + indexName + "] onto" +
-                    " prior to shrink"));
-            }
+                } else {
+                    // No nodes currently match the allocation rules or have no enough disk bytes,
+                    // so report this as an error and we'll retry
+                    logger.debug("could not find any nodes to allocate index [{}] onto prior to shrink", indexName);
+                    listener.onFailure(new NoNodeAvailableException("could not find any nodes to allocate index [" + indexName +
+                        "] onto prior to shrink"));
+                }
+            }, (Exception e) -> listener.onFailure(new NoNodeAvailableException("failed to retrieve disk information" +
+                " to select a single node for shard copy allocation"))));
+
         } else {
             // There are no shards for the index, the index might be gone. Even though this is a retryable step ILM will not retry in
             // this case as we're using the periodic loop to trigger the retries and that is run over *existing* indices.
             listener.onFailure(new IndexNotFoundException(indexMetadata.getIndex()));
         }
     }
 
+    Optional<String> selectSingleNode(Set<String> validNodeIds, Map<String, Long> nodeShardsStorageBytes) {
```

Comment:
can you add some docs for this please?

Reply:
I've add some description in the ilm-shrink doc.